### PR TITLE
Add packages and configuration changes to allow Luna to start in QEMU

### PR DESCRIPTION
**Pull request by magmastonealex**
Luna will now start in QEMU. To do so (either on this branch, or once everything's in master):
- Install postmarketos-ui-luna
- Run /usr/bin/start_luna.sh
- The first-use app should appear. Complete the app (Click on where the "next" button should be for the step after license agreements that just has a black screen).
- Run `luna-send -n 1 'luna://com.palm.configurator/run' '{"types":["dbkinds","dbpermissions","activities"]}'` in another terminal
- Stop luna, and reboot.
- Run /usr/bin/start_luna.sh

Much of the race conditions that require the reboots, manual `configurator` running, and long delays in `start_luna.sh` can be dealt with using openRC startup files.

Most apps appear to work, with the notable exception of the email client (which is missing services that it needs).

ARM support is still broken because of #1008 and #878. 

(more work towards #629)

### CODE REVIEW

File: aports/luna/enyo/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+pkgdesc="EnyoJS webapp framework"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/enyo-1.0/archive/$_commit.tar.gz"
+builddir="$srcdir/$pkgname-1.0-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/enyo/APKBUILD]: Empty `build()` functions can be removed.

File: aports/luna/luna-app-services/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/app-services/archive/$_commit.tar.gz"
+options="!check"
+builddir="$srcdir/app-services-$_commit"
+
+prepare() {
+	default_prepare
+	mkdir -p "$srcdir"/build
+}
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/luna-app-services/APKBUILD]: empty build function

File: aports/luna/luna-app-browser/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+package() {
+	cd "$srcdir"/build
+	make DESTDIR="$pkgdir" install
+
+	install -d "$pkgdir"/etc/palm/db/kinds
+	install -d "$pkgdir"/etc/palm/db/permissions
+	install -v -m644 "$builddir"/configuration/db/kinds/* "$pkgdir"/etc/palm/db/kinds
+	install -v -m644 "$builddir"/configuration/db/permissions/* "$pkgdir"/etc/palm/db/permissions
+
```
Comment (ollieparanoid) [inline aports/luna/luna-app-browser/APKBUILD]: unnecessary empty lines

File: aports/luna/luna-app-services/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	install -d "$pkgdir"/usr/share/ls2-dev/roles/pub
+	install -d "$pkgdir"/usr/share/ls2-dev/roles/prv
+	install -d "$pkgdir"/var/palm/data/com.palm.appInstallService
+	install -d "$pkgdir"/usr/palm/public/accounts
+	install -d "$pkgdir"/etc/palm/tempdb/kinds
+	install -d "$pkgdir"/etc/palm/tempdb/permissions
+	install -d "$pkgdir"/etc/palm/event.d
+	install -d "$pkgdir"/etc/palm/init
+
+    for SERVICE in `ls -d1 ./com.palm.service.*` ; do
```
Comment (ollieparanoid) [inline aports/luna/luna-app-services/APKBUILD]: That syntax is obsolete as far as I know, the new one would be:
```shell
for SERVICE in $(ls -d1 ./com.palm.service.*); do
```

File: aports/luna/luna-app-services/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        SERVICE_DIR=`basename $SERVICE`
```
Comment (ollieparanoid) [inline aports/luna/luna-app-services/APKBUILD]: see above

**Comment by montvid** [comment]
For testing purposes i installed Luna on a armhf Nexus 7 2013 (it has mesa support - hardware acceleration) here is the error log - reminds me of #878. I also needed to install x11 it seems to get no warnings/errors:

apk add xorg-server xf86-input-libinput libinput
https://paste.gnome.org/poiclprj6

File: aports/luna/luna-app-services/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	cp -vrf com.palm.service.accounts/tempdb/kinds/* "$pkgdir"/etc/palm/tempdb/kinds/
+	cp -vrf com.palm.service.accounts/tempdb/permissions/* "$pkgdir"/etc/palm/tempdb/permissions
+
+	# install account service upstart files
+	install -m 644 com.palm.service.accounts/files/etc/event.d/createLocalAccount "$pkgdir"/etc/palm/event.d/
+	install -m 644 com.palm.service.accounts/files/etc/init/createLocalAccount.conf "$pkgdir"/etc/palm/init/
+
+	# create folder for contact linker plugins
+	mkdir -p "$pkgdir"/etc/palm/contact_linker_plugins
+
```
Comment (ollieparanoid) [inline aports/luna/luna-app-services/APKBUILD]: empty lines

File: aports/luna/luna-applauncher/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+pkgdesc="WebOS app launcher"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends="enyo webos-underscore mojoloader webos-foundation"
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/$pkgname/archive/$_commit.tar.gz"
+builddir="$srcdir/$pkgname-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/luna-applauncher/APKBUILD]: empty build function

**Comment by magmastonealex** [comment]
@montvid Correct, #878 means that you can't use Luna on armhf devices yet. You shouldn't need those x11 packages - there are some warnings without them, but everything should still work.

File: aports/luna/luna-core-apps/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+pkgdesc="The new webOS UI of the webOS ports project."
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends="luna-webappmanager enyo luna-app-services"
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/core-apps/archive/$_commit.tar.gz"
+builddir="$srcdir/core-apps-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/luna-core-apps/APKBUILD]: empty build

File: aports/luna/luna-core-apps/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	return 0;
+}
+
+package() {
+	
+	cd "$builddir"
+
+    install -d "$pkgdir"/usr/palm/applications
```
Comment (ollieparanoid) [inline aports/luna/luna-core-apps/APKBUILD]: spaces instead of tabs

File: aports/luna/luna-core-apps/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        if [ -d $COREAPPS/configuration/db/permissions ]; then
+            install -v -m 644 $COREAPPS/configuration/db/permissions/* "$pkgdir"/etc/palm/db/permissions
+        fi
+
+        if [ -d $COREAPPS/configuration/activities ]; then
+            cp -vrf $COREAPPS/configuration/activities/* "$pkgdir"/etc/palm/activities/
+        fi
+    done
+
+    # TODO: Upstream removes these because they package enyo-2 based versions as well.
```
Comment (ollieparanoid) [inline aports/luna/luna-core-apps/APKBUILD]: How about opening an issue instead, and remove the whole block with the commented out code, including this TODO message? Personally I am not a fan of commented out code.

File: aports/luna/luna-firstuse/APKBUILD [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+		-DCMAKE_BUILD_TYPE=Release \
+		-DCMAKE_INSTALL_PREFIX=/usr \
+		-DWEBOS_INSTALL_ROOT=/
+	make
+}
+
+package() {
+	cd "$srcdir"/build
+	make DESTDIR="$pkgdir" install
+
```
Comment (ollieparanoid) [inline aports/luna/luna-firstuse/APKBUILD]: empty line

File: aports/luna/luna-init/APKBUILD [hunk 12]
```diff
@@ -28,6 +28,16 @@ build() {
 
 package() {
 	cd "$srcdir"/build
+	#The cmake install file doesn't install everything in the right spot.
 	make DESTDIR="$pkgdir" install
+
+	install -d "$pkgdir"/etc/palm
+    install -v -m 644 "$builddir"/files/conf/locale.txt "$pkgdir"/etc/palm
```
Comment (ollieparanoid) [inline aports/luna/luna-init/APKBUILD]: spaces

File: aports/luna/luna-systemui/APKBUILD [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+pkgdesc="WebOS power menu and other system ui"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends="enyo"
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/$pkgname/archive/$_commit.tar.gz"
+builddir="$srcdir/$pkgname-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/luna-systemui/APKBUILD]: empty build

File: aports/luna/luna-temp-ofono-bluetooth/APKBUILD [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+pkgdesc="Temporary package for ofono/bluetooth mocking."
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="luneos-components-$_commit.tar.gz::https://github.com/webOS-ports/luneos-components/archive/$_commit.tar.gz"
+builddir="$srcdir/luneos-components-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/luna-temp-ofono-bluetooth/APKBUILD]: empty build

File: aports/luna/mojoloader/APKBUILD [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+pkgdesc="mojoloader"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/mojoloader/archive/$_commit.tar.gz"
+builddir="$srcdir/$pkgname-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/mojoloader/APKBUILD]: empty build

File: aports/luna/mojoloader/APKBUILD [hunk 16]
```diff
@@ -0,0 +1,27 @@
+pkgname=mojoloader
+pkgver=1.1_git20161005
+pkgrel=0
+_commit=4eec67e2c941ab416aaced46c604c2ff3747c2fd
+pkgdesc="mojoloader"
```
Comment (ollieparanoid) [inline aports/luna/mojoloader/APKBUILD]: What does this program do?

File: aports/luna/mojoloader/APKBUILD [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/mojoloader/archive/$_commit.tar.gz"
+builddir="$srcdir/$pkgname-$_commit"
+
+build() {
+	return 0
+}
+
+package() {
+
```
Comment (ollieparanoid) [inline aports/luna/mojoloader/APKBUILD]: empty lines

File: aports/luna/mojoservice-frameworks/APKBUILD [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+pkgdesc="Webos core services javascript frameworks"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webos-ports/mojoservice-frameworks/archive/$_commit.tar.gz"
+builddir="$srcdir/mojoservice-frameworks-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/mojoservice-frameworks/APKBUILD]: empty build

File: aports/luna/mojoservice-frameworks/APKBUILD [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+	return 0
+}
+
+package() {
+	cd "$builddir"
+
+	install -d "$pkgdir"/usr/palm/frameworks
+
+	for FRAMEWORK in `ls -d1 ./mojoservice*` ; do
```
Comment (ollieparanoid) [inline aports/luna/mojoservice-frameworks/APKBUILD]: `$()`?

File: aports/luna/nodejs-module-webos-pmlog/APKBUILD [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+pkgver=18.0.0_git20150424
+pkgrel=0
+_commit=ca359ed90389c5be085e7eac0548236312743d6a
+pkgdesc="Library to allow nodejs code to talk to palm pmloglib."
+arch="all"
+url="https://github.com/webOS-ports/nodejs-module-webos-pmlog"
+license="GPL-3.0+"
+depends="pmloglib nodejs nodejs-npm"
+
+#vim is in makedepends because it provides xxd
```
Comment (ollieparanoid) [inline aports/luna/nodejs-module-webos-pmlog/APKBUILD]: missing space after "#"

File: aports/luna/webos-foundation/APKBUILD [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+pkgdesc="Webos webapp javascript frameworks"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webos-ports/foundation-frameworks/archive/$_commit.tar.gz"
+builddir="$srcdir/foundation-frameworks-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/webos-foundation/APKBUILD]: empty build

File: aports/luna/webos-loadable-frameworks/APKBUILD [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+pkgdesc="Webos core services javascript frameworks"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/webos-ports/loadable-frameworks/archive/$_commit.tar.gz"
+builddir="$srcdir/loadable-frameworks-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/webos-loadable-frameworks/APKBUILD]: empty build

File: aports/luna/webos-underscore/APKBUILD [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+pkgdesc="webos packaged version of underscore.js"
+arch="all"
+url="http://webos-ports.org"
+license="GPL-3.0+"
+depends=""
+makedepends=""
+source="$pkgname-$_commit.tar.gz::https://github.com/openwebos/underscore/archive/$_commit.tar.gz"
+builddir="$srcdir/underscore-$_commit"
+
+build() {
```
Comment (ollieparanoid) [inline aports/luna/webos-underscore/APKBUILD]: empty build

File: aports/main/postmarketos-ui-luna/APKBUILD [hunk 24]
```diff
@@ -1,11 +1,11 @@
 pkgname=postmarketos-ui-luna
 pkgver=1
-pkgrel=0
+pkgrel=1
 pkgdesc="(Wayland) webOS UI, ported from the LuneOS project (Not working yet)"
 url="https://wiki.postmarketos.org/wiki/LuneOS_UI"
 arch="noarch"
 license="GPL3+"
-depends="luna-next-cardshell luna-next luna-sysmgr luna-sysservice luna-appmanager luneos-components luna-init mesa"
+depends="luna-next-cardshell luna-next luna-sysmgr luna-sysservice luna-appmanager luneos-components luna-init mesa db8 luna-webappmanager configurator activitymanager enyo luna-firstuse luna-systemui luna-qml-launcher luna-applauncher mojoloader luna-temp-ofono-bluetooth luna-app-services xkeyboard-config luna-core-apps luna-app-browser"
```
Comment (ollieparanoid) [inline aports/main/postmarketos-ui-luna/APKBUILD]: I recommend line breaks here. You could also put one package per line there, and sort them alphabetically.

File: aports/main/postmarketos-ui-luna/start_luna.sh [hunk 25]
```diff
@@ -1,27 +1,36 @@
...
+LunaSysService -l debug &
+sleep 1
+export QT_QPA_PLATFORM=eglfs
+export QT_QPA_EGLFS_DISABLE_INPUT=1
+luna-next -plugin evdevmouse:/dev/input/event2 -plugin evdevkeyboard:/dev/input/event1 &
+sleep 5
+export QT_QPA_PLATFORM=wayland
+LunaAppManager -t -c -u luna &
+sleep 1
+LunaWebAppManager --verbose --allow-file-access-from-files
```
Comment (ollieparanoid) [inline aports/main/postmarketos-ui-luna/start_luna.sh]: I know that this is just a workaround right now.
Do the daemons write anything to stdout or stderr? If so, I would recommend to redirect them to the syslog.

File: aports/luna/luna-app-browser/APKBUILD [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+license="GPL-3.0+"
+depends="qt5-qtquickcontrols2 qt5-qtquickcontrols qt5-qtwebengine"
+makedepends="cmake-modules-webos qt5-qtquickcontrols2-dev qt5-qtwebengine-dev"
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/org.webosports.app.browser/archive/$_commit.tar.gz
+fix-qtwebengine-private-includes.patch"
+builddir="$srcdir/org.webosports.app.browser-$_commit"
+
+prepare() {
+	default_prepare
+	#cmake does not allow in-source builds
```
Comment (ollieparanoid) [inline aports/luna/luna-app-browser/APKBUILD]: missing space after "#"

File: aports/luna/luna-firstuse/APKBUILD [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+license="GPL-3.0+"
+depends="qt5-qtquickcontrols2 qt5-qtquickcontrols"
+makedepends="cmake-modules-webos qt5-qtquickcontrols2-dev"
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/org.webosports.app.firstuse/archive/$_commit.tar.gz
+locale.patch"
+builddir="$srcdir/org.webosports.app.firstuse-$_commit"
+
+prepare() {
+	default_prepare
+	#cmake does not allow in-source builds
```
Comment (ollieparanoid) [inline aports/luna/luna-firstuse/APKBUILD]: missing space after "#"

File: aports/luna/luna-init/APKBUILD [hunk 28]
```diff
@@ -28,6 +28,16 @@ build() {
 
 package() {
 	cd "$srcdir"/build
+	#The cmake install file doesn't install everything in the right spot.
```
Comment (ollieparanoid) [inline aports/luna/luna-init/APKBUILD]: missing space after "#"

File: aports/luna/luna-webappmanager/APKBUILD [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+makedepends="cmake-modules-webos glib-dev luna-service2-dev libpbnjson-dev luna-sysmgr-common-dev luna-prefs-dev qt5-qtbase-dev qt5-qtquickcontrols2-dev qt5-qtdeclarative-dev qt5-qtwebengine-dev qt5-qtwebchannel-dev libconnman-qt-dev"
+source="$pkgname-$_commit.tar.gz::https://github.com/webOS-ports/luna-webappmanager/archive/$_commit.tar.gz
+remove-systemd.patch
+fix-qtwebengine-private-includes.patch
+tryFix.patch"
+builddir="$srcdir/$pkgname-$_commit"
+
+prepare() {
+	default_prepare
+	#cmake does not allow in-source builds
```
Comment (ollieparanoid) [inline aports/luna/luna-webappmanager/APKBUILD]: missing space after "#"

**Comment by magmastonealex** [comment]
@ollieparanoid Unfortunately yes, `/usr/palm` is scattered all over the code, it'd be quite tricky to try and patch it out everywhere.

I think I've addressed everything else you brought up, and I removed all of the prepare statements that I added in this PR

File: aports/main/postmarketos-ui-luna/start_luna.sh (hunk 25 above)
Comment (ollieparanoid) [reply aports/main/postmarketos-ui-luna/start_luna.sh]: @magmastonealex: thanks for being so fast!
This is the only thing left unanswered.

Comment (magmastonealex) [reply aports/main/postmarketos-ui-luna/start_luna.sh]: Ah, OK. I piped all of the ones with output to `logger` with tags, and verified they show up in `logread`.

**Comment by montvid** [comment]
I just installed it from pmbootstrap menu. I can try again after this is merged.

**Comment by montvid** [comment]
Oh @ollieparanoid disabled it for armhf?

**Comment by magmastonealex** [comment]
@montvid, yes, disabled because qtwebengine doesn't build for armhf right now due to #1008. That means that Luna can't be used.

It makes sense you didn't have any success from the menu, you were missing everything that would have made it work :)

**Pull request merged**